compiler/src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -17,112 +17,93 @@ import scala.annotation.internal.sharable
 
 object TyperState {
   @sharable private var nextId: Int = 0
+  def initialState() =
+    TyperState()
+      .init(null, OrderingConstraint.empty)
+      .setReporter(new ConsoleReporter())
+      .setCommittable(true)
 }
 
-class TyperState(private val previous: TyperState /* | Null */) {
+class TyperState() {
 
-  Stats.record("typerState")
+  private var myId: Int = _
+  def id: Int = myId
 
-  val id: Int = TyperState.nextId
-  TyperState.nextId += 1
+  private var previous: TyperState = _
```

Comment:
Maybe worth keeping the hint that this can legitimately be initialized to null:
```suggestion
  private var previous: TyperState /* | Null */ = _
```